### DISCUSSION TITLE
Built-in script minify support?

### DISCUSSION THREAD

**Discussion by aliencaocao** [pyinstaller/pyinstaller#6353] [Features & Development]
I am actually not sure if this is already a feature, but if it is not, I think it will come in handy for some people distributing small-sized apps.

I found out this project https://python-minifier.com/ that is able to minify any python script without affecting its functionality. They do provide a way to call it within a python script (see https://dflook.github.io/python-minifier/api_usage.html), so I wonder if PyInstaller could add an optional switch to enable this and run every python script dependency through this. It may bring down the size of smaller apps by a lot since it removes docstrings and blank lines, something that many mature libraries are stuffed with.

Meanwhile, I can of course minify my own application script, but I can't do anything to the dependencies (libraries) used since I would not know exactly all python files my app depends on, while PyInstaller does this task well.

I am using UPX, but not sure whether it strips all the extra stuff from the source code that python-minify does.

**Comment by bwoodsend** [comment]
This will have much less of an impact than you think because PyInstaller ships compiled bytecode (.pyc files) instead of source code (.py files) and Python has put a lot of work into making bytecode very compact. Going through the list of pyminfier optimisations:

#### Combine Import statements

Multiple imports on one line is a feature of the source code only and is normalised to one per line by the compiler. Combining imports has no effect on bytecode.

```python
>>> compile("import foo\nimport bar", "<>", "exec").co_code
b'd\x00d\x01l\x00Z\x00d\x00d\x01l\x01Z\x01d\x01S\x00'
>>> compile("import foo, bar", "<>", "exec").co_code
b'd\x00d\x01l\x00Z\x00d\x00d\x01l\x01Z\x01d\x01S\x00'  # <- this is identical to the 1st
```

#### Remove Pass statements

Both using `pass` and `0` are no-ops and are removed by the compiler. Hence replacing one with the other has no effect on the bytecode or its size.

#### Remove literal statements (docstrings)

This does make a small difference although it's a bit of a dangerous one because sometimes you do encounter libraries which either use their docstrings for something other than documentation (e.g. anything using [PLY](https://www.dabeaz.com/ply/)) or do some dynamic docstring generation which fails if those doctrings disappear (NumPy used to be bad for this). This functionality can already be achieved by using `python -OO -m PyInstaller some-code.py` although [we really don't recommend it](https://github.com/pyinstaller/pyinstaller/issues/6345).

#### Evaluate Literal Escapes 

Again, no difference to bytecode because any nonliteral sequence of characters in a source code literal is replaced with its literal value.
```python
>>> compile('a = "hello\u5675"', "<>", "exec").co_consts
('hello噵', None)
```

#### Hoist Literals

The compiler already coalesces duplicate literals.

```python
>>> compile('a = "hello"\nb = "hello"\nif a == "hello": pass', "<>", "exec").co_consts
('hello', None)  # <-- only one copy of the "hello" string
```
On the flip side, pyminifier's storing of duplicate literals in global variables actually makes the size worse because it has added extra statements.

```python
>>> len(marshal.dumps(compile('a = "hello"\nb = "hello"\nif a == "hello": pass', "<>", "exec")))
107  # <- size without optimisation
>>> len(marshal.dumps(compile("_A='hello'\na=_A\nb=_A\nif a==_A:0", "<>", "exec")))
117  # <- size with optimisation is greater than without!
```

#### Remove Annotations

This will also make a small difference but it is also a dangerous one. Packages like [beartype](https://github.com/beartype/beartype) use those type hints for type checking.

#### Rename Locals/Globals

This does make a difference - it's much smaller than you'd expect because, like with literals, each unique variable name is only written once in bytecode. This would make me want to cry if I ever had to debug an application via either `pdb` or `code.interact()` but is otherwise OK.

#### Convert Positional-Only Arguments to Normal Arguments

This makes a very small difference.


It's also worth pointing out that PyInstaller puts all the bytecode in a zlib compressed archive so that for every 100 bytes you save in raw bytecode you will only see an overall reduction off about 20-40 bytes in the overall compressed archive. Source code really isn't a significant factor in program size.

**Reply by aliencaocao** [reply]
Thanks for the detailed analysis! Now I am convinced that using minify is too much of a trouble. However, I got one question. You said pyinstaller compiles source code into bytecode, but when I manually check the temp extraction path in the sys.MEIPASS path, I can recover my full script in .py with all the comments and stuff. Is that supposed to happen? From what I understand, if the executable somehow recovered the original source code in its full format, it is not really compressing it, right?

**Reply by rokm** [reply]
Some packages (`tensorflow`, `torch`) *do* require source .py files to be collected, so if you use any of those (or if you use custom hooks for packages with `collect_all` calls), you end up with source code collected for such packages.

**Reply by aliencaocao** [reply]
Well for me it's actually my own application script, like main,py. I did not use tensorflow. The only non-built in packages used are pandas and Bleak

**Reply by rokm** [reply]
The entry-point script is not collected as source by default. Can we see your spec file?

**Reply by aliencaocao** [reply]
So I just did another test run and the main.py is no longer there. Last time I saw it being collected is when I used 4.5.1 and some other libs, but now I am using 5.0dev0 and removed those libs. I guess this is no longer a concern for me now. Thanks for the help!
